shared/domains/chelonia/internals.js

```diff
@@ -158,47 +162,220 @@ export default (sbp('sbp/selectors/register', {
       }
     }
   },
+  'chelonia/private/out/latestHash': function (contractID: string) {
+    return fetch(`${this.config.connectionURL}/latestHash/${contractID}`, {
+      cache: 'no-store'
+    }).then(handleFetchResult('text'))
+  },
+  // TODO: r.body is a stream.Transform, should we use a callback to process
+  //       the events one-by-one instead of converting to giant json object?
+  //       however, note if we do that they would be processed in reverse...
+  'chelonia/private/out/eventsSince': async function (contractID: string, since: string) {
+    const events = await fetch(`${this.config.connectionURL}/eventsSince/${contractID}/${since}`)
+      .then(handleFetchResult('json'))
+    if (Array.isArray(events)) {
+      return events.reverse().map(b64ToStr)
+    }
+  },
   'chelonia/private/in/processMessage': async function (message: GIMessage, state: Object) {
     const [opT, opV] = message.op()
     const hash = message.hash()
     const contractID = message.contractID()
     const manifestHash = message.manifest()
     const config = this.config
-    if (!state._vm) state._vm = {}
+    const signature = message.signature()
+    const signedPayload = message.signedPayload()
+    const env = this.env
+    const self = this
+    if (!state._vm) state._vm = Object.create(null)
     const opFns: { [GIOpType]: (any) => void } = {
       [GIMessage.OP_CONTRACT] (v: GIOpContract) {
-        // TODO: shouldn't each contract have its own set of authorized keys?
-        if (!state._vm.authorizedKeys) state._vm.authorizedKeys = []
-        // TODO: we probably want to be pushing the de-JSON-ified key here
-        state._vm.authorizedKeys.push({ key: v.keyJSON, context: 'owner' })
+        const keys = { ...env.additionalKeys, ...state._volatile?.keys }
+        state._vm.authorizedKeys = keysToMap(v.keys)
+
+        for (const key of v.keys) {
+          if (key.meta?.private) {
+            if (key.id && key.meta.private.keyId in keys && key.meta.private.content) {
+              if (!state._volatile) state._volatile = Object.create(null)
+              if (!state._volatile.keys) state._volatile.keys = Object.create(null)
```

Comment:
Just curious - is there a safe way to combine these two checks into one? I.e. if we could use `{}` instead of `Object.create(null)` then this could've been 1-line:

```js
if (!state._volatile) state._volatile = { keys: {} }
```

Reply:
`{}` isn't used because of prototype pollution, but the reason to have two checks is that `state._volatile` could be defined but not have a `keys` property. So, the second check is still necessary somehow.

Perhaps `if (!state._volatile?.keys) state._volatile = {...state._volatile, keys: state._volatile?.keys ?? {}}` would work as a 'single' check (and I could change it to something like that), but the two checks are still being done, just implicitly and in a less efficient manner.